File: flow-client/src/main/java/com/vaadin/client/DefaultRegistry.java
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.client;

import java.util.function.Supplier;

import com.google.gwt.xhr.client.XMLHttpRequest;

import com.vaadin.client.communication.AtmospherePushConnection;
import com.vaadin.client.communication.AtmospherePushConnectionCallbacks;
import com.vaadin.client.communication.DefaultConnectionStateHandler;
import com.vaadin.client.communication.DefaultConnectionStateHandlerCallbacks;
import com.vaadin.client.communication.Heartbeat;
import com.vaadin.client.communication.HeartbeatCallbacks;
import com.vaadin.client.communication.LoadingIndicatorStateHandler;
import com.vaadin.client.communication.MessageHandler;
import com.vaadin.client.communication.MessageHandlerCallbacks;
import com.vaadin.client.communication.MessageSender;
import com.vaadin.client.communication.MessageSenderCallbacks;
import com.vaadin.client.communication.Poller;
import com.vaadin.client.communication.PushConfiguration;
import com.vaadin.client.communication.ReconnectConfiguration;
import com.vaadin.client.communication.RequestResponseTracker;
import com.vaadin.client.communication.ServerConnector;
import com.vaadin.client.communication.ServerRpcQueue;
import com.vaadin.client.communication.XhrConnection;
import com.vaadin.client.communication.XhrConnectionCallbacks;
import com.vaadin.client.flow.ConstantPool;
import com.vaadin.client.flow.ExecuteJavaScriptProcessor;
import com.vaadin.client.flow.StateTree;
import com.vaadin.flow.shared.ApplicationConstants;
import com.vaadin.flow.shared.util.SharedUtil;

/**
 * A registry implementation used by {@link ApplicationConnection}.
 *
 * @author Vaadin Ltd
 * @since 1.0
 */
public class DefaultRegistry extends Registry {

    /**
     * Constructs a registry based on the given application connection and
     * configuration references.
     *
     * @param connection
     *            the application connection
     * @param applicationConfiguration
     *            the application configuration
     */
    public DefaultRegistry(ApplicationConnection connection,
            ApplicationConfiguration applicationConfiguration) {
        // Note that initialization order matters. Many constructors depend on
        // ApplicationConnection, ApplicationConfiguration and StateTree even
        // though this is not explicitly specified anywhere.

        set(ApplicationConnection.class, connection);
        set("ApplicationConfiguration", applicationConfiguration);

        // Classes with no constructor dependencies
        ResourceLoader resourceLoader = new ResourceLoader(
                message -> getSystemErrorHandler().handleError(message), true);
        set("ResourceLoader", resourceLoader);
        URIResolver uriResolver = new URIResolver(this);
        set("URIResolver", uriResolver);
        set("DependencyLoader",
                new DependencyLoader(uriResolver, resourceLoader));
        set("UILifecycle", (Supplier<UILifecycle>) UILifecycle::new);
        UILifecycle uiLifecycle = get("UILifecycle");
        StateTree stateTree = new StateTree(this);
        set("StateTree", stateTree);
        RequestResponseTracker requestResponseTracker = new RequestResponseTracker(
                () -> {
                    MessageSender ms = getMessageSender();
                    if ((getUILifecycle().isRunning()
                            && getServerRpcQueue().isFlushPending())
                            || ms.getResynchronizationState() == MessageSender.ResynchronizationState.SEND_TO_SERVER
                            || ms.hasQueuedMessages()) {
                        ms.sendInvocationsToServer();
                    }
                });
        set("RequestResponseTracker", requestResponseTracker);

        MessageHandlerCallbacks messageHandlerCallbacks = new MessageHandlerCallbacks();
        messageHandlerCallbacks.setGetMessageSender(this::getMessageSender);
        messageHandlerCallbacks.setGetUiLifecycle(this::getUILifecycle);
        messageHandlerCallbacks.setGetStateTree(this::getStateTree);
        messageHandlerCallbacks.setGetConstantPool(this::getConstantPool);
        messageHandlerCallbacks
                .setGetSystemErrorHandler(this::getSystemErrorHandler);
        messageHandlerCallbacks.setGetExecuteJavaScriptProcessor(
                this::getExecuteJavaScriptProcessor);
        messageHandlerCallbacks
                .setGetDependencyLoader(this::getDependencyLoader);
        messageHandlerCallbacks.setGetResourceLoader(this::getResourceLoader);
        messageHandlerCallbacks
                .setGetRequestResponseTracker(this::getRequestResponseTracker);
        messageHandlerCallbacks.setGetLoadingIndicatorStateHandler(
                this::getLoadingIndicatorStateHandler);
        messageHandlerCallbacks.setGetApplicationConfiguration(
                this::getApplicationConfiguration);
        messageHandlerCallbacks.setRedirect(WidgetUtil::redirect);
        set("MessageHandler", new MessageHandler(messageHandlerCallbacks));

        MessageSenderCallbacks messageSenderCallbacks = new MessageSenderCallbacks();
        messageSenderCallbacks.setGetMessageHandler(this::getMessageHandler);
        messageSenderCallbacks.setGetUiLifecycle(this::getUILifecycle);
        messageSenderCallbacks
                .setGetRequestResponseTracker(this::getRequestResponseTracker);
        messageSenderCallbacks.setGetLoadingIndicatorStateHandler(
                this::getLoadingIndicatorStateHandler);
        messageSenderCallbacks
                .setGetPushConfiguration(this::getPushConfiguration);
        messageSenderCallbacks.setGetServerRpcQueue(this::getServerRpcQueue);
        messageSenderCallbacks.setGetApplicationConfiguration(
                this::getApplicationConfiguration);
        messageSenderCallbacks
                .setSendXhr(payload -> getXhrConnection().send(payload));
        messageSenderCallbacks.setGetXhrUri(() -> getXhrConnection().getUri());
        messageSenderCallbacks
                .setCreatePushConnection(this::createPushConnection);
        MessageSender messageSender = new MessageSender(messageSenderCallbacks);
        set("MessageSender", messageSender);

        ServerRpcQueue serverRpcQueue = new ServerRpcQueue(uiLifecycle,
                () -> messageSender.sendInvocationsToServer());
        set("ServerRpcQueue", serverRpcQueue);
        LoadingIndicatorStateHandler loadingIndicatorStateHandler = new LoadingIndicatorStateHandler(
                requestResponseTracker::hasActiveRequest);
        set("LoadingIndicatorStateHandler", loadingIndicatorStateHandler);
        ServerConnector serverConnector = new ServerConnector(
                loadingIndicatorStateHandler, serverRpcQueue);
        set("ServerConnector", serverConnector);
        stateTree.setServerConnector(serverConnector);
        set("ExecuteJavaScriptProcessor", new ExecuteJavaScriptProcessor(this));
        set("ConstantPool", (Supplier<ConstantPool>) ConstantPool::new);
        set("ExistingElementMap",
                (Supplier<ExistingElementMap>) ExistingElementMap::new);
        InitialPropertiesHandler initialPropertiesHandler = new InitialPropertiesHandler(
                stateTree);
        set("InitialPropertiesHandler", initialPropertiesHandler);
        stateTree.setInitialPropertiesHandler(initialPropertiesHandler);

        // Classes with dependencies, in correct order
        Supplier<Heartbeat> heartbeatSupplier = () -> {
            String uri = SharedUtil.addGetParameter(
                    applicationConfiguration.getServiceUrl(),
                    ApplicationConstants.REQUEST_TYPE_PARAMETER,
                    ApplicationConstants.REQUEST_TYPE_HEARTBEAT);
            uri = SharedUtil.addGetParameter(uri,
                    ApplicationConstants.UI_ID_PARAMETER,
                    applicationConfiguration.getUIId());
            HeartbeatCallbacks callbacks = new HeartbeatCallbacks() {
                @Override
                public void onOk() {
                    getConnectionStateHandler().heartbeatOk();
                }

                @Override
                public void onInvalidStatusCode(XMLHttpRequest xhr) {
                    getConnectionStateHandler().heartbeatInvalidStatusCode(xhr);
                }

                @Override
                public void onException(XMLHttpRequest xhr, String message) {
                    getConnectionStateHandler().heartbeatException(xhr,
                            message);
                }
            };
            return new Heartbeat(uri,
                    applicationConfiguration.getHeartbeatInterval(),
                    uiLifecycle, callbacks);
        };
        set(Heartbeat.class, heartbeatSupplier);

        DefaultConnectionStateHandlerCallbacks connectionStateCallbacks = new DefaultConnectionStateHandlerCallbacks();
        connectionStateCallbacks.setGetUiLifecycle(this::getUILifecycle);
        connectionStateCallbacks
                .setGetSystemErrorHandler(this::getSystemErrorHandler);
        connectionStateCallbacks.setGetHeartbeat(this::getHeartbeat);
        connectionStateCallbacks
                .setGetReconnectConfiguration(this::getReconnectConfiguration);
        connectionStateCallbacks
                .setGetRequestResponseTracker(this::getRequestResponseTracker);
        connectionStateCallbacks.setGetLoadingIndicatorStateHandler(
                this::getLoadingIndicatorStateHandler);
        connectionStateCallbacks.setGetApplicationConfiguration(
                this::getApplicationConfiguration);
        connectionStateCallbacks.setGetMessageSender(this::getMessageSender);
        set("ConnectionStateHandler",
                new DefaultConnectionStateHandler(connectionStateCallbacks));
        XhrConnectionCallbacks xhrConnectionCallbacks = new XhrConnectionCallbacks();
        xhrConnectionCallbacks.setGetMessageHandler(this::getMessageHandler);
        xhrConnectionCallbacks
                .setGetConnectionStateHandler(this::getConnectionStateHandler);
        xhrConnectionCallbacks
                .setGetRequestResponseTracker(this::getRequestResponseTracker);
        xhrConnectionCallbacks.setGetApplicationConfiguration(
                this::getApplicationConfiguration);
        set("XhrConnection", new XhrConnection(xhrConnectionCallbacks));
        set("PushConfiguration",
                new PushConfiguration(stateTree,
                        () -> messageSender.setPushEnabled(true),
                        () -> messageSender.setPushEnabled(false)));
        set("ReconnectConfiguration", new ReconnectConfiguration(stateTree));
        set("Poller", new Poller(stateTree, uiLifecycle));

        // Wire SystemErrorHandler last so its callbacks can reach the rest of
        // the registry through `this::get*` lookups without forward references.
        SystemErrorHandlerCallbacks systemErrorCallbacks = new SystemErrorHandlerCallbacks() {
            @Override
            public String getServiceUrl() {
                return applicationConfiguration.getServiceUrl();
            }

            @Override
            public boolean isWebComponentMode() {
                return applicationConfiguration.isWebComponentMode();
            }

            @Override
            public boolean isProductionMode() {
                return applicationConfiguration.isProductionMode();
            }

            @Override
            public com.vaadin.client.bootstrap.ErrorMessage getSessionExpiredError() {
                return applicationConfiguration.getSessionExpiredError();
            }

            @Override
            public String[] getExportedWebComponents() {
                return applicationConfiguration.getExportedWebComponents();
            }

            @Override
            public int getHeartbeatInterval() {
                return applicationConfiguration.getHeartbeatInterval();
            }

            @Override
            public void setHeartbeatInterval(int seconds) {
                getHeartbeat().setInterval(seconds);
            }

            @Override
            public boolean isPushEnabled() {
                return getPushConfiguration().isPushEnabled();
            }

            @Override
            public void setPushEnabled(boolean enabled) {
                getMessageSender().setPushEnabled(enabled);
            }

            @Override
            public void disablePushImmediately() {
                getMessageSender().setPushEnabled(false, false);
            }

            @Override
            public void applyResyncResponse(String responseText) {
                int uiId = applicationConfiguration.getUIId();
                com.vaadin.client.ValueMap json = MessageHandler
                        .parseJson(responseText);
                int newUiId = json.getInt(ApplicationConstants.UI_ID);
                if (newUiId != uiId) {
                    com.vaadin.client.Console
                            .debug("UI ID switched from " + uiId + " to "
                                    + newUiId + " after resynchronization");
                    applicationConfiguration.setUIId(newUiId);
                }
                reset();
                getUILifecycle().setState(UILifecycle.UIState.RUNNING);
                getMessageHandler().handleMessage(json);
            }
        };
        // Keyed by an explicit string because the Registry uses
        // Class.getName() as a key by default; GWT collapses every native
        // @JsType to JavaScriptObject so multiple such registrations would
        // collide (e.g. with the Heartbeat registration above).
        set("SystemErrorHandler", new SystemErrorHandler(systemErrorCallbacks));
    }

    /**
     * Builds the callbacks struct for a new {@link AtmospherePushConnection}
     * and returns the freshly constructed instance. Used by
     * {@link MessageSender} (through the {@code createPushConnection} callback)
     * so the push connection is only instantiated when push is actually
     * enabled.
     */
    private AtmospherePushConnection createPushConnection() {
        AtmospherePushConnectionCallbacks pushCallbacks = new AtmospherePushConnectionCallbacks();
        pushCallbacks.setGetUiLifecycle(this::getUILifecycle);
        pushCallbacks.setGetPushConfiguration(this::getPushConfiguration);
        pushCallbacks.setGetApplicationConfiguration(
                this::getApplicationConfiguration);
        pushCallbacks.setGetURIResolver(this::getURIResolver);
        pushCallbacks.setGetMessageHandler(this::getMessageHandler);
        pushCallbacks
                .setGetConnectionStateHandler(this::getConnectionStateHandler);
        pushCallbacks.setGetResourceLoader(this::getResourceLoader);
        return new AtmospherePushConnection(pushCallbacks);
    }

}

File: flow-client/src/main/frontend/internal/DefaultRegistry.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

// TypeScript port of com.vaadin.client.DefaultRegistry — the concrete service
// assembly used by ApplicationConnection. It instantiates the ported TS services
// in dependency order (initialization order matters: later constructors read
// earlier services via the getters) and exposes the typed getters every service
// resolves its collaborators through. This is the cutover-assembly step that
// takes the build-alongside TS engine into use.
//
// Push is wired XHR-only for now: MessageSender is created without a
// PushConnectionFactory, so setPushEnabled(true) is a no-op until the full
// AtmospherePushConnection class is wired in as the factory.

import type { ApplicationConfiguration } from './ApplicationConfiguration';
import { ConstantPool } from './ConstantPool';
import { DependencyLoader } from './DependencyLoader';
import { DefaultConnectionStateHandler } from './communication/DefaultConnectionStateHandler';
import { Heartbeat } from './communication/Heartbeat';
import { LoadingIndicatorStateHandler } from './communication/LoadingIndicatorStateHandler';
import { MessageHandler } from './MessageHandler';
import { MessageSender } from './communication/MessageSender';
import { Poller } from './communication/Poller';
import { PushConfiguration } from './communication/PushConfiguration';
import { ReconnectConfiguration } from './communication/ReconnectConfiguration';
import { Registry } from './Registry';
import { RequestResponseTracker } from './communication/RequestResponseTracker';
import { ResourceLoader } from './ResourceLoader';
import { ServerConnector } from './communication/ServerConnector';
import { ServerRpcQueue } from './communication/ServerRpcQueue';
import { ExecuteJavaScriptProcessor } from './ExecuteJavaScriptProcessor';
import { ExistingElementMap } from './ExistingElementMap';
import { InitialPropertiesHandler } from './InitialPropertiesHandler';
import { StateTree } from './StateTree';
import { SystemErrorHandler } from './SystemErrorHandler';
import { UILifecycle } from './UILifecycle';
import { URIResolver } from './URIResolver';
import { XhrConnection } from './XhrConnection';

// Service lookup tokens (one per registered singleton).
const TOKEN = {
  ApplicationConfiguration: 'ApplicationConfiguration',
  ResourceLoader: 'ResourceLoader',
  URIResolver: 'URIResolver',
  DependencyLoader: 'DependencyLoader',
  SystemErrorHandler: 'SystemErrorHandler',
  UILifecycle: 'UILifecycle',
  StateTree: 'StateTree',
  RequestResponseTracker: 'RequestResponseTracker',
  MessageHandler: 'MessageHandler',
  MessageSender: 'MessageSender',
  ServerRpcQueue: 'ServerRpcQueue',
  ServerConnector: 'ServerConnector',
  ExecuteJavaScriptProcessor: 'ExecuteJavaScriptProcessor',
  ConstantPool: 'ConstantPool',
  ExistingElementMap: 'ExistingElementMap',
  InitialPropertiesHandler: 'InitialPropertiesHandler',
  Heartbeat: 'Heartbeat',
  ConnectionStateHandler: 'ConnectionStateHandler',
  XhrConnection: 'XhrConnection',
  PushConfiguration: 'PushConfiguration',
  ReconnectConfiguration: 'ReconnectConfiguration',
  Poller: 'Poller',
  LoadingIndicatorStateHandler: 'LoadingIndicatorStateHandler'
} as const;

/** The concrete registry that wires the TS services; mirrors DefaultRegistry.java. */
export class DefaultRegistry extends Registry {
  constructor(applicationConfiguration: ApplicationConfiguration) {
    super();
    // Initialization order matters: many constructors read earlier services.
    const self = this as never;

    this.set(TOKEN.ApplicationConfiguration, applicationConfiguration);

    // No constructor dependencies (resolve collaborators lazily via getters).
    this.set(
      TOKEN.ResourceLoader,
      // ResourceLoader takes an error handler directly; adapt it to resolve the
      // SystemErrorHandler lazily (it is registered just below).
      new ResourceLoader({ handleError: (message: string) => this.getSystemErrorHandler().handleError(message) }, true)
    );
    this.set(TOKEN.URIResolver, new URIResolver(self));
    this.set(TOKEN.DependencyLoader, new DependencyLoader(self));
    this.set(TOKEN.SystemErrorHandler, new SystemErrorHandler(self));
    this.setResettable(TOKEN.UILifecycle, () => new UILifecycle());
    this.set(TOKEN.StateTree, new StateTree(self));
    this.set(TOKEN.RequestResponseTracker, new RequestResponseTracker(self));
    this.set(TOKEN.MessageHandler, new MessageHandler(self));
    this.set(TOKEN.MessageSender, new MessageSender(self));
    this.set(TOKEN.ServerRpcQueue, new ServerRpcQueue(self));
    this.set(TOKEN.ServerConnector, new ServerConnector(self));
    this.set(TOKEN.ExecuteJavaScriptProcessor, new ExecuteJavaScriptProcessor(self));
    this.setResettable(TOKEN.ConstantPool, () => new ConstantPool());
    this.setResettable(TOKEN.ExistingElementMap, () => new ExistingElementMap());
    this.set(TOKEN.InitialPropertiesHandler, new InitialPropertiesHandler(self));

    // Classes with dependencies, in order.
    this.setResettable(TOKEN.Heartbeat, () => new Heartbeat(self));
    this.set(TOKEN.ConnectionStateHandler, new DefaultConnectionStateHandler(self));
    this.set(TOKEN.XhrConnection, new XhrConnection(self));
    this.set(TOKEN.PushConfiguration, new PushConfiguration(self));
    this.set(TOKEN.ReconnectConfiguration, new ReconnectConfiguration(self));
    this.set(TOKEN.Poller, new Poller(self));
    this.set(TOKEN.LoadingIndicatorStateHandler, new LoadingIndicatorStateHandler(self));
  }

  getApplicationConfiguration(): ApplicationConfiguration {
    return this.get(TOKEN.ApplicationConfiguration);
  }

  getResourceLoader(): ResourceLoader {
    return this.get(TOKEN.ResourceLoader);
  }

  getURIResolver(): URIResolver {
    return this.get(TOKEN.URIResolver);
  }

  getDependencyLoader(): DependencyLoader {
    return this.get(TOKEN.DependencyLoader);
  }

  getSystemErrorHandler(): SystemErrorHandler {
    return this.get(TOKEN.SystemErrorHandler);
  }

  getUILifecycle(): UILifecycle {
    return this.get(TOKEN.UILifecycle);
  }

  getStateTree(): StateTree {
    return this.get(TOKEN.StateTree);
  }

  getRequestResponseTracker(): RequestResponseTracker {
    return this.get(TOKEN.RequestResponseTracker);
  }

  getMessageHandler(): MessageHandler {
    return this.get(TOKEN.MessageHandler);
  }

  getMessageSender(): MessageSender {
    return this.get(TOKEN.MessageSender);
  }

  getServerRpcQueue(): ServerRpcQueue {
    return this.get(TOKEN.ServerRpcQueue);
  }

  getServerConnector(): ServerConnector {
    return this.get(TOKEN.ServerConnector);
  }

  getExecuteJavaScriptProcessor(): ExecuteJavaScriptProcessor {
    return this.get(TOKEN.ExecuteJavaScriptProcessor);
  }

  getConstantPool(): ConstantPool {
    return this.get(TOKEN.ConstantPool);
  }

  getExistingElementMap(): ExistingElementMap {
    return this.get(TOKEN.ExistingElementMap);
  }

  getInitialPropertiesHandler(): InitialPropertiesHandler {
    return this.get(TOKEN.InitialPropertiesHandler);
  }

  getHeartbeat(): Heartbeat {
    return this.get(TOKEN.Heartbeat);
  }

  getConnectionStateHandler(): DefaultConnectionStateHandler {
    return this.get(TOKEN.ConnectionStateHandler);
  }

  getXhrConnection(): XhrConnection {
    return this.get(TOKEN.XhrConnection);
  }

  getPushConfiguration(): PushConfiguration {
    return this.get(TOKEN.PushConfiguration);
  }

  getReconnectConfiguration(): ReconnectConfiguration {
    return this.get(TOKEN.ReconnectConfiguration);
  }

  getPoller(): Poller {
    return this.get(TOKEN.Poller);
  }

  getLoadingIndicatorStateHandler(): LoadingIndicatorStateHandler {
    return this.get(TOKEN.LoadingIndicatorStateHandler);
  }
}
